### PR TITLE
firrtl memory generator node lowering

### PR DESCRIPTION
**Pull request by darthscsi**
Lower firrtl memory generator nodes into a simulation model.  This implements the code in LowerToRTL as a separate pass in preparation for converging lowering to generate generator modules.

### CODE REVIEW

File: lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      op->getAttrOfType<IntegerAttr>("numReadWritePorts").getUInt();
+  mem.readLatency = op->getAttrOfType<IntegerAttr>("readLatency").getUInt();
+  mem.writeLatency = op->getAttrOfType<IntegerAttr>("writeLatency").getUInt();
+  mem.dataWidth = op->getAttrOfType<IntegerAttr>("width").getUInt();
+  mem.readUnderWrite =
+      op->getAttrOfType<IntegerAttr>("readUnderWrite").getUInt();
+  return mem;
+};
+
+static Value addPipelineStages(ImplicitLocOpBuilder &b, size_t stages,
```
Comment (mikeurbach) [inline lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp:60]: Nothing to do about this, but for the record, this seems like another place we could use #587.

**Review by mikeurbach — COMMENTED** [review]
I'm not in a place to approve this, but I took a quick glance and it makes sense. Thanks for adding this. I will test it out and add any feedback here.

File: lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp:60]: Yup, though I was please with how compact it wound up being.

Comment (darthscsi) [reply lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp:60]: It's doing what scala firrtl is doing and generating a register per value per delay cycle, which is ugly, but can be cleaned up later.

Comment (mikeurbach) [reply lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp:60]: Yep totally agree.

**Review by mikeurbach — APPROVED** [review]
Aside from one edge case, which I mentioned over there, this works for me with the latest of #864.

**Review by seldridge — APPROVED** [review]
Awesome. This is a nice clean lowering.

I'm not sure how to handle the readwrite latency situation. Using the max makes sense, but maybe this should be a verifier error. I'm asking around for clarification. Not blocking this PR.

File: lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    // Read Logic
+    Value ren =
+        b.create<sv::ReadInOutOp>(b.create<sv::ArrayIndexInOutOp>(reg, addr));
+    Value x = b.create<sv::ConstantXOp>(dataType);
+
+    Value rdata = b.create<comb::MuxOp>(en, ren, x);
+    outputs.push_back(rdata);
+  }
+  for (size_t i = 0; i < mem.numReadWritePorts; ++i) {
+    auto numStages = std::max(mem.readLatency, mem.writeLatency) - 1;
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp:108]: yes, though having RW ports allowing reads to bypass writes is totally bogus.  For cases where the rw ports should be split, I would expect that to happen upstream of this pass.

File: test/Dialect/SV/rtl-memsim.mlir [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+//CHECK-NEXT:    %[[wcond:.+]] = comb.and %wo_en_0, %wo_mask_0
+//CHECK-NEXT:    sv.if %[[wcond]]  {
+//CHECK-NEXT:      %[[wslot:.+]] = sv.array_index_inout %Memory[%wo_addr_0]
+//CHECK-NEXT:      sv.passign %[[wslot]], %wo_data_0
+//CHECK-NEXT:    }
+//CHECK-NEXT:  }
+//CHECK-NEXT:  rtl.output %[[readres]], %[[rwres]]
+
+//CHECK-LABEL: @FIRRTLMem_1_1_1_16_10_2_4_0
+//CHECK:     %Memory = sv.reg  : !rtl.inout<uarray<10xi16>>
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply test/Dialect/SV/rtl-memsim.mlir]: Yea, checking one field of one port is tractable.

**Pull request merged**